File: backend/app/services/meter_service.py
"""Orchestrates meter creation/lookup: validates input, then delegates to the
meter repository for persistence.
"""

from datetime import date, datetime, timezone

from sqlmodel import Session

from app.core.logging import get_logger
from app.domain.errors import NotFoundError
from app.models.enums import SolarMode
from app.models.meter import Meter
from app.repositories import meter_repository

logger = get_logger(__name__)


def create_meter(
    session: Session,
    *,
    user_id: str,
    meter_number: str,
    display_name: str | None = None,
    solar_mode: SolarMode = SolarMode.NONE,
    billing_cycle_reference_date: date | None = None,
    cycle_length_months: int = 2,
) -> Meter:
    if not meter_number or not meter_number.strip():
        raise ValueError("Meter number is required.")
    if cycle_length_months <= 0:
        raise ValueError("cycle_length_months must be a positive number of months.")

    meter = Meter(
        user_id=user_id,
        meter_number=meter_number.strip(),
        display_name=display_name,
        solar_mode=solar_mode,
        billing_cycle_reference_date=billing_cycle_reference_date,
        cycle_length_months=cycle_length_months,
    )
    created = meter_repository.create(session, meter)
    logger.info("Created meter %s (meter_number=%s)", created.id, created.meter_number)
    return created


def get_meter(session: Session, meter_id: str, user_id: str) -> Meter:
    """Raises NotFoundError both when the meter doesn't exist at all and
    when it belongs to a different user - the caller shouldn't be able to
    tell those two cases apart (CLAUDE.md section 14: authorization checks
    on every household/meter resource).
    """
    meter = meter_repository.get(session, meter_id)
    if meter is None or meter.user_id != user_id:
        raise NotFoundError(f"Meter {meter_id} not found.")
    return meter


def list_meters(session: Session, user_id: str, active_only: bool = False) -> list[Meter]:
    return meter_repository.list_all(session, user_id, active_only=active_only)


def update_meter(
    session: Session,
    meter_id: str,
    user_id: str,
    *,
    display_name: str | None = None,
    solar_mode: SolarMode | None = None,
    active: bool | None = None,
    billing_cycle_reference_date: date | None = None,
    cycle_length_months: int | None = None,
    last_assessment_date: date | None = None,
    next_expected_assessment_date: date | None = None,
) -> Meter:
    meter = get_meter(session, meter_id, user_id)

    if cycle_length_months is not None:
        if cycle_length_months <= 0:
            raise ValueError("cycle_length_months must be a positive number of months.")
        meter.cycle_length_months = cycle_length_months
    if display_name is not None:
        meter.display_name = display_name
    if solar_mode is not None:
        meter.solar_mode = solar_mode
    if active is not None:
        meter.active = active
    if billing_cycle_reference_date is not None:
        meter.billing_cycle_reference_date = billing_cycle_reference_date
    if last_assessment_date is not None:
        meter.last_assessment_date = last_assessment_date
    if next_expected_assessment_date is not None:
        meter.next_expected_assessment_date = next_expected_assessment_date

    meter.updated_at = datetime.now(timezone.utc)
    updated = meter_repository.update(session, meter)
    logger.info("Updated meter %s", meter_id)
    return updated

File: frontend/lib/services/meter_service.dart
import '../core/api_client.dart';
import '../core/json_helpers.dart';
import '../models/meter.dart';
import '../models/solar_mode.dart';

/// Talks to the backend's /meters endpoints. Mirrors backend/app/api/meters.py.
class MeterApiService {
  final ApiClient client;

  MeterApiService(this.client);

  Future<List<Meter>> listMeters({bool activeOnly = false}) async {
    final json = await client.get('/meters', query: {'active_only': activeOnly.toString()});
    return (json as List<dynamic>).map((e) => Meter.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Meter> getMeter(String id) async {
    final json = await client.get('/meters/$id');
    return Meter.fromJson(json as Map<String, dynamic>);
  }

  Future<Meter> createMeter({
    required String meterNumber,
    String? displayName,
    SolarMode solarMode = SolarMode.none,
    DateTime? billingCycleReferenceDate,
    int cycleLengthMonths = 2,
  }) async {
    final json = await client.post('/meters', {
      'meter_number': meterNumber,
      'display_name': displayName,
      'solar_mode': solarMode.apiValue,
      'billing_cycle_reference_date':
          billingCycleReferenceDate == null ? null : formatDateForApi(billingCycleReferenceDate),
      'cycle_length_months': cycleLengthMonths,
    });
    return Meter.fromJson(json as Map<String, dynamic>);
  }

  Future<Meter> updateMeter(
    String id, {
    String? displayName,
    SolarMode? solarMode,
    bool? active,
    DateTime? billingCycleReferenceDate,
    int? cycleLengthMonths,
    DateTime? lastAssessmentDate,
    DateTime? nextExpectedAssessmentDate,
  }) async {
    final body = <String, dynamic>{};
    if (displayName != null) body['display_name'] = displayName;
    if (solarMode != null) body['solar_mode'] = solarMode.apiValue;
    if (active != null) body['active'] = active;
    if (billingCycleReferenceDate != null) {
      body['billing_cycle_reference_date'] = formatDateForApi(billingCycleReferenceDate);
    }
    if (cycleLengthMonths != null) body['cycle_length_months'] = cycleLengthMonths;
    if (lastAssessmentDate != null) body['last_assessment_date'] = formatDateForApi(lastAssessmentDate);
    if (nextExpectedAssessmentDate != null) {
      body['next_expected_assessment_date'] = formatDateForApi(nextExpectedAssessmentDate);
    }

    final json = await client.patch('/meters/$id', body);
    return Meter.fromJson(json as Map<String, dynamic>);
  }
}
